libmamba/tests/src/util/test_graph.cpp

```diff
@@ -339,11 +375,139 @@ TEST_SUITE("graph")
     {
         DiGraph<int> g;
         test_visitor<DiGraph<int>> vis;
-        g.depth_first_search(vis);
+        dfs_raw(g, vis, 0);
         CHECK(vis.get_back_edge_map().empty());
         CHECK(vis.get_cross_edge_map().empty());
     }
 
+    template <typename Graph, typename Iter>
+    auto is_node_id_permutation(const Graph& g, Iter first, Iter last)->bool
+    {
+        using node_id = typename Graph::node_id;
+        auto node_ids = std::vector<node_id>();
+        g.for_each_node_id([&node_ids](node_id n) { node_ids.push_back(n); });
+        return std::is_permutation(first, last, node_ids.cbegin(), node_ids.cend());
+    }
+
+    TEST_CASE("dfs_all")
+    {
+        DiGraph<double> g;
+        const auto n0 = g.add_node(0);
+        const auto n1 = g.add_node(1);
+        const auto n2 = g.add_node(2);
+        g.add_edge(n0, n1);
+        g.add_edge(n2, n1);
+
+        test_visitor<decltype(g)> vis = {};
+        dfs_raw(g, vis);
+
+
+        const auto& start_node_list = vis.get_start_node_list();
+        CHECK(is_node_id_permutation(g, start_node_list.cbegin(), start_node_list.cend()));
+        const auto& finish_node_list = vis.get_finish_node_list();
+        CHECK(is_node_id_permutation(g, finish_node_list.cbegin(), finish_node_list.cend()));
+        const auto start_node_set = std::set(start_node_list.begin(), start_node_list.end());
+        const auto finish_node_set = std::set(finish_node_list.begin(), finish_node_list.end());
+        CHECK_EQ(start_node_set, finish_node_set);
+        CHECK_EQ(start_node_set.size(), 3);
+    }
+
+    TEST_CASE("dfs_preorder & dfs_postorder")
+    {
+        DiGraph<double> g;
+        const auto n0 = g.add_node(0);
+        const auto n1 = g.add_node(1);
+        const auto n2 = g.add_node(2);
+        g.add_edge(n0, n1);
+        g.add_edge(n2, n1);
+
+        using node_id = typename decltype(g)::node_id;
+        auto nodes = std::vector<node_id>();
+
+        SUBCASE("dfs_preorder starting on a given node")
+        {
+            dfs_preorder_nodes_for_each_id(
+                g,
+                [&nodes](node_id n) { nodes.push_back(n); },
+                n0
+            );
+            CHECK_EQ(nodes, std::vector<node_id>{ n0, n1 });
+        }
+
+        SUBCASE("dfs_preorder on all nodes")
+        {
+            REQUIRE(g.has_node(n0));
+            REQUIRE(g.has_node(n1));
+            REQUIRE(g.has_node(n2));
+            dfs_preorder_nodes_for_each_id(g, [&nodes](node_id n) { nodes.push_back(n); });
+            CHECK(is_node_id_permutation(g, nodes.cbegin(), nodes.cend()));
+            CHECK_EQ(nodes, std::vector<node_id>{ n0, n1, n2 });
```

Comment:
IIUC shouldn't `n2` come first (before `n1`)? Since `n1` is `n2`'s child?

Reply:
No that's desired outcome. The dfs on all node starts with n0, finds n1, and stops. Then it starts again with the next unvisited node, n2, and finishes as n1 is already visited.

Reply:
Even in preorder? Ok!

Reply:
Yes, a better way to think of it it's that is is like two calls to `dfs` (as dfs does not explore everything on the first run).